x-pack/plugins/reporting/server/lib/content_stream.ts

```diff
@@ -249,8 +251,52 @@ export class ContentStream extends Duplex {
     });
   }
 
-  private async flush(size = this.buffer.byteLength) {
-    const chunk = this.buffer.slice(0, size);
+  private async flush(size = this.bytesBuffered) {
+    const buffersToFlush: Buffer[] = [];
+    let partiallyFlushedBuffer: undefined | Buffer;
+    let bytesToFlush = 0;
+
+    /*
+     Loop over each buffer, keeping track of how many bytes we have added
+     to the array of buffers to be flushed. The array of buffers to be flushed
+     contains buffers by reference, not copies. This avoids putting pressure on
+     the CPU for copying buffers or for gc activity. Please profile performance
+     with a large byte configuration and a large number of records (900k+)
+     before changing this code. Config used at time of writing:
+
+      xpack.reporting:
+        csv.maxSizeBytes: 500000000
+        csv.scroll.size: 1000
+
+     At the moment this can put memory pressure on Kibana. Up to 1,1 GB in a dev
+     build. It is not recommended to have overly large max size bytes but we
+     need this code to be as performant as possible.
+    */
+    for (const buffer of this.buffers) {
+      const remainder = size - bytesToFlush;
+      if (remainder < 0) {
+        break;
+      }
+      const chunkedBuffer = remainder > 0 ? buffer.slice(0, remainder) : buffer;
```

Comment:
It seems like an error -- we are adding the whole buffer if the `remainder` is zero.
Isn't `slice` expensive too?

```suggestion
      const chunkedBuffer = remainder >= buffer.byteLength ? buffer : buffer.slice(0, remainder);
```

Reply:
I'll take a look, slice is not an expensive call AFAICT: https://nodejs.org/api/buffer.html#bufslicestart-end

It's the same underlying memory just a different view on it.

Reply:
If it is not an expensive call, then it can be simplified:
```suggestion
      const chunkedBuffer = buffer.slice(0, remainder);
```